planner/optimize.go

```diff
@@ -46,28 +45,58 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 
 	sctx.PrepareTxnFuture(ctx)
 
-	var oriHint *bindinfo.HintsSet
-	if sctx.GetSessionVars().UsePlanBaselines {
-		if stmtNode, ok := node.(ast.StmtNode); ok {
-			oriHint = addHint(sctx, stmtNode)
-		}
+	bestPlan, names, _, err := optimize(ctx, sctx, node, is)
+	if err != nil || !sctx.GetSessionVars().UsePlanBaselines {
+		return bestPlan, names, err
+	}
+	stmtNode, ok := node.(ast.StmtNode)
+	if !ok {
+		return bestPlan, names, nil
 	}
-	plan, names, err := optimize(ctx, sctx, node, is)
-	// Restore the original hint in case of prepare stmt.
-	if oriHint != nil {
-		node = bindinfo.BindHint(node.(ast.StmtNode), oriHint)
+	bindRecord, scope := getBindRecord(sctx, stmtNode)
+	if bindRecord == nil {
+		return bestPlan, names, nil
+	}
+	bestPlanHint := plannercore.GenHintsFromPhysicalPlan(bestPlan)
+	bestCostAmongHints := math.MaxFloat64
+	var bestPlanAmongHints plannercore.Plan
+	originHints := bindinfo.CollectHint(stmtNode)
+	// Try to find the best binding.
+	for _, binding := range bindRecord.Bindings {
+		if binding.Status != bindinfo.Using {
+			continue
+		}
+		metrics.BindUsageCounter.WithLabelValues(scope).Inc()
+		bindinfo.BindHint(stmtNode, binding.Hint)
+		plan, _, cost, err := optimize(ctx, sctx, node, is)
 		if err != nil {
-			handleInvalidBindRecord(ctx, sctx, node.(ast.StmtNode))
+			binding.Status = bindinfo.Invalid
+			handleInvalidBindRecord(sctx, scope, bindinfo.BindRecord{
+				OriginalSQL: bindRecord.OriginalSQL,
+				Db:          bindRecord.Db,
+				Bindings:    []bindinfo.Binding{binding},
+			})
+			continue
+		}
+		// The best bestPlan is in baselines, just use it.
+		if bestPlanHint == plannercore.GenHintsFromPhysicalPlan(plan) {
```

Comment:
Can we just compare `bestPlanHint` with hints in `bindRecord`, and avoid the `optimize` for these bindings if it is found in the list?

Reply:
Good point.